be2-scala/src/main/scala/ch/epfl/pop/model/network/method/message/data/DataBuilder.scala

```diff
@@ -66,9 +80,26 @@ object DataBuilder {
     case ActionType.WITNESS => WitnessMessage.buildFromJson(payload)
     case _ => throw new ProtocolException(s"Unknown action '$action' encountered while creating a Witness Data")
   }
-
+  
   private def buildSocialMediaData(action: ActionType, payload: String): MessageData = action match {
     case ActionType.ADD => AddChirp.buildFromJson(payload)
     case _ => throw new ProtocolException(s"Unknown action '$action' encountered while creating a Social Media Data")
   }
+  
+  /**
+    * Builds a message payload after passing a schema validation check
+    *
+    * @param payload payload to build
+    * @param validator one of the validators at [[DataSchemaValidator]] to valid the schema of the payload
+    * @param buildFromJson the data builder
+    * @param errMsg error message to include in description in case of error
+    * @return built MessageData or throws an exceptition in case of schema failure
+    */
+  @throws(classOf[ProtocolException])
+  private def buildOrReject(payload: String)(validator: String => Try[Unit])(buildFromJson: String => MessageData )(errMsg: String): MessageData = {
+      validator(payload) match {
+        case Success(_) => buildFromJson(payload)
+        case Failure(e) => throw new ProtocolException(errMsg + ": " + e.getMessage)
```

Comment:
maybe for coherence use ProtocolException(s"$errMsg: ${e.getMessage}") as above

Reply:
👍